src/triage/component/architect/protected_groups_generators.py

```diff
@@ -0,0 +1,125 @@
+import logging
+import textwrap
+from triage.database_reflection import table_exists
+
+DEFAULT_PROTECTED_GROUPS_NAME = "protected_groups"
+
+
+class ProtectedGroupsGeneratorNoOp(object):
+    def generate_all_dates(self):
+        logging.warning(
+            "No bias audit configuration is available, so protected groups will not be created"
+        )
+
+    def generate(self):
+        logging.warning(
+            "No bias audit configuration is available, so protected groups will not be created"
+        )
+
+    def clean_up(self, protected_groups_table_name):
+        pass
+
+
+class ProtectedGroupsGenerator(object):
+    def __init__(self, db_engine, from_obj, attribute_columns, entity_id_column, knowledge_date_column, protected_groups_table_name=None, replace=True):
+        self.db_engine = db_engine
+        self.replace = replace
+        self.protected_groups_table_name = protected_groups_table_name if protected_groups_table_name else DEFAULT_PROTECTED_GROUPS_NAME
+        self.from_obj = from_obj
+        self.attribute_columns = attribute_columns
+        self.entity_id_column = entity_id_column
+        self.knowledge_date_column = knowledge_date_column
+
+    def _create_protected_groups_table(self):
+        if self.replace or not table_exists(self.protected_groups_table_name, self.db_engine):
+            self.db_engine.execute("drop table if exists {}".format(self.protected_groups_table_name))
+            self.db_engine.execute(
+                """
+                create table {} (
+                entity_id int,
+                as_of_date date,
+                {}
+
+            )""".format(
+                    self.protected_groups_table_name,
+                    ", ".join([str(col) + " varchar(60)" for col in self.attribute_columns])
+                )
+            )
+        else:
+            logging.info("Not dropping and recreating protected groups table because "
+                         "replace flag was set to False and table was found to exist")
+
+    def generate_all_dates(self,  as_of_dates,  cohort_table_name):
+        self._create_protected_groups_table()
+        logging.info(
+            "Creating protected_groups for %s as of dates and %s label timespans",
+            len(as_of_dates)
+        )
+        for as_of_date in as_of_dates:
+            if not self.replace:
+                logging.info(
+                    "Looking for existing protected_groups for as of date %s",
+                    as_of_date
+                )
+                any_existing_rows = list(self.db_engine.execute(
+                    """select 1 from {protected_groups_table}
+                    where as_of_date = '{as_of_date}'
+                    limit 1
+                    """.format(
+                        protected_groups_table=self.protected_groups_table_name,
+                        as_of_date=as_of_date
+                    )
+                ))
+                if len(any_existing_rows) == 1:
```

Comment:
I believe `scalar()` could clean up the two queries here and below, _e.g._:

```
any_existing_rows = self.db_engine.execute('''select 1 from …''').scalar()
if any_existing_rows:  # 1 or None
    …
```

[[docs]](https://kite.com/python/docs/sqlalchemy.engine.result.ResultProxy.scalar)

I'd also recommend `f"…"` strings for many of these, for the same (minor) reason.